File: 1383.最大的团队表现值.cpp
#include <vector>
#include <utility>
#include <queue>

using std::vector;
using std::priority_queue;


class Solution {
public:
    // customized sort & priority queue
    int maxPerformance(int n, vector<int>& speed, vector<int>& efficiency, int k) {
        vector<std::pair<int, int>> vec(n);
        for (int i = 0; i < n; ++i) {
            vec[i] = {efficiency[i], speed[i]};
        }
        std::sort(rbegin(vec), rend(vec));
        long sum = 0, ans = 0;
        // simulate a less-than queue
        priority_queue<int, vector<int>, std::greater<>> pq;
        for (auto& [e, s] : vec) {
            pq.emplace(s);
            sum += s;
            if (pq.size() > k) {
                sum -= pq.top();
                pq.pop();
            }
            ans = std::max(ans, sum * e);
        }
        return ans % int(1e9 + 7);
    }
};

File: 1383.最大的团队表现值.py

import heapq
from typing import List

class Solution:
    # priority queue
    def maxPerformance(self, n: int, speed: List[int], efficiency: List[int], k: int) -> int:
        vec = []
        ans = ssum = 0
        # customized sort: efficiency descending
        for e, s in sorted(zip(efficiency, speed), reverse=1):
            # simulate a less-than priority queue of speed
            heapq.heappush(vec, s)
            ssum += s
            if len(vec) > k:
                ssum -= heapq.heappop(vec)
            ans = max(ans, ssum * e)
        return ans % (10 ** 9 + 7)
